Add utility methods to find an upgrader

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68358
* #69915
* #69914
* __->__ #68355

Differential Revision: [D33198156](https://our.internmc.facebook.com/intern/diff/D33198156)

tugsbayasgalan committed:
Add utility methods to find an upgrader

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add utility methods to find an upgrader"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add utility methods to find an upgrader"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add utility methods to find an upgrader"




[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/operator_upgraders/upgraders.h:
```diff
@@ -0,0 +1,76 @@
+namespace torch {
+namespace jit {
+
+static std::unordered_map<std::string, std::string> upgraders_graph(
```
Is it temporary to have IR text in the file? When will we be able to compile python upgraders on the fly during `torch.jit.load`?

tugsbayasgalan replied:
https://github.com/pytorch/pytorch/pull/68773 This is the stack of PRs that will make sure the upgraders are generated on the fly. This stack will be landed after that.

gmagogsfm commented on torch/csrc/jit/operator_upgraders/utils.h:
```diff
@@ -0,0 +1,55 @@
+  return true;
+}
+
+static bool isOpSymbolCurrent(std::string name, int current_version) {
```
Ping.

tugsbayasgalan replied:
Resolved now

cccclai commented on torch/csrc/jit/operator_upgraders/version_map.h:
```diff
@@ -0,0 +1,30 @@
+    {{"aten::div.Tensor",
+      {{4,
+        "div_Tensor_0_3",
+        "aten::div.Tensor(Tensor self, Tensor other) -> Tensor"}}},
```
Why old schema is needed here?

tugsbayasgalan replied:
I need it to compile the older models with their specific schema and replace it with the upgraders later.

cccclai commented on torch/csrc/jit/operator_upgraders/utils.h:
```diff
@@ -0,0 +1,55 @@
+      upgraders_for_schema.begin(),
+      upgraders_for_schema.end(),
+      [current_version](const UpgraderEntry& entry) {
+        return entry.version_bump > current_version;
```
My understanding is that for `div_Tensor_0_3`, the loaded model operator version needs to be within [0,3]

tugsbayasgalan replied:
I thought it was more like versions between 0 and 3 need to be upgraded with the current upgrader. So div_Tensor_0_3 implies there was a version bump which is greater than 3. @gmagogsfm What do you think about this check?

tugsbayasgalan committed:
Update on "Add utility methods to find an upgrader"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add utility methods to find an upgrader"




[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/operator_upgraders/utils.h:
```diff
@@ -0,0 +1,55 @@
+    if (entry.version_bump > current_version) {
+      return false;
+    }
+
```
Nit:
this empty line breaks formatting

tugsbayasgalan replied:
resolved

tugsbayasgalan committed:
Update on "Add utility methods to find an upgrader"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add utility methods to find an upgrader"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add utility methods to find an upgrader"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add utility methods to find an upgrader"




[ghstack-poisoned]

tugsbayasgalan commented on torch/csrc/jit/operator_upgraders/utils.h:
```diff
@@ -0,0 +1,55 @@
+}
+
+static bool isOpEntryCurrent(
+    std::vector<UpgraderEntry> upgraders_for_schema,
```
resolved

tugsbayasgalan commented on torch/csrc/jit/operator_upgraders/utils.h:
```diff
@@ -0,0 +1,55 @@
+namespace torch {
+namespace jit {
+
+static c10::optional<UpgraderEntry> findUpgrader(
```
changed to inline

tugsbayasgalan replied:
Moved all function definitions to seperate cpp file

tugsbayasgalan committed:
Update on "Add utility methods to find an upgrader"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Add utility methods to find an upgrader"




[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198156).

tugsbayasgalan committed:
Update on "Add utility methods to find an upgrader"


Differential Revision: [D33198156](https://our.internmc.facebook.com/intern/diff/D33198156)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198156).

tugsbayasgalan committed:
Update on "Add utility methods to find an upgrader"


Differential Revision: [D33198156](https://our.internmc.facebook.com/intern/diff/D33198156)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198156).

gmagogsfm commented on torch/csrc/jit/operator_upgraders/utils.cpp:
```diff
@@ -0,0 +1,53 @@
+  //    1. the version entry must be greater than current_version
+  //    2. Among the version entries, we need to see if the current version
+  //       is in the upgrader name range
+  auto pos = std::find_if(
+      upgraders_for_schema.begin(),
+      upgraders_for_schema.end(),
+      [current_version](const UpgraderEntry& entry) {
+        return entry.bumped_at_version > current_version;
+      });
```
This assumes upgrader entries are sorted by `bumped_at_version`. Is this guaranteed?

tugsbayasgalan replied:
Yes. We can explitcitly mention it should be sorted. Since the version map is used frequently, I feel like it would be expensive to keep sorting it. But it is probably better if the developers maintain the sorted order

cccclai replied:
If developers are expected to sort the order, it will be good to add a test to prevent adding unsorted entry.

tugsbayasgalan replied:
I think a unittest makes more sense. Because we add values to this map only for testing purposes.

tugsbayasgalan replied:
Now we have unittest, and also a sorting mechanism

gmagogsfm commented on torch/csrc/jit/operator_upgraders/utils.cpp:
```diff
@@ -0,0 +1,53 @@
+  return c10::nullopt;
+}
+
+bool isOpEntryCurrent(
```
What is "OpEntry"? And what does it mean for an "OpEntry" to be "Current"? IIRC, based on our last discussion, only ops can be current.

tugsbayasgalan replied:
Yeah this is a bad naming. Will update

tugsbayasgalan replied:
done

gmagogsfm commented on torch/csrc/jit/operator_upgraders/utils.cpp:
```diff
@@ -0,0 +1,53 @@
+    const std::vector<UpgraderEntry>& upgraders_for_schema,
+    size_t current_version) {
+  for (const auto& entry : upgraders_for_schema) {
+    if (entry.bumped_at_version > current_version) {
```
If upgraders are sorted, we can simply check the last entry in vector to know whether if an upgrader is needed for op.

tugsbayasgalan replied:
Yep will do.

tugsbayasgalan replied:
done

gmagogsfm commented on torch/csrc/jit/operator_upgraders/utils.h:
```diff
@@ -0,0 +1,26 @@
+namespace torch {
+namespace jit {
+
+TORCH_API c10::optional<UpgraderEntry> findUpgrader(
+    const std::vector<UpgraderEntry>& upgraders_for_schema,
+    size_t current_version);
+
+TORCH_API bool isOpEntryCurrent(
+    const std::vector<UpgraderEntry>& upgraders_for_schema,
+    size_t current_version);
+
+TORCH_API bool isOpSymbolCurrent(
+    const std::string& name,
+    size_t current_version);
```
Consider adding comments to describe what they do.

tugsbayasgalan replied:
sounds good

tugsbayasgalan replied:
done

gmagogsfm commented on test/cpp/jit/test_upgrader_utils.cpp:
```diff
@@ -0,0 +1,22 @@
+
+namespace torch {
+namespace jit {
+
```
Could we add test cases for other utilities like `isOpCurrent`?

tugsbayasgalan replied:
sounds good

tugsbayasgalan replied:
done

gmagogsfm commented on test/cpp/jit/test_upgrader_utils.cpp:
```diff
@@ -0,0 +1,22 @@
+      {8, "foo__4_7", "foo.bar()"},
+  };
+
+  auto upgrader = findUpgrader(dummy_entry, 6);
```
Consider testing other cases too, like `1`, `10`

tugsbayasgalan replied:
yep.

tugsbayasgalan replied:
done

tugsbayasgalan committed:
Update on "Add utility methods to find an upgrader"


Differential Revision: [D33198156](https://our.internmc.facebook.com/intern/diff/D33198156)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198156).

tugsbayasgalan committed:
Update on "Add utility methods to find an upgrader"


Differential Revision: [D33198156](https://our.internmc.facebook.com/intern/diff/D33198156)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198156).

qihqi commented on torch/csrc/jit/operator_upgraders/version_map.cpp:
```diff
@@ -0,0 +1,57 @@
+  return operatorVersionMap;
+}
+
+void test_only_add_entry(std::string op_name, UpgraderEntry entry) {
```
const std::string&

tugsbayasgalan replied:
done

qihqi commented on torch/csrc/jit/operator_upgraders/version_map.cpp:
```diff
@@ -0,0 +1,57 @@
+  operatorVersionMap[op_name].push_back(entry);
+}
+
+void test_only_remove_entry(std::string op_name) {
```
const std::string&

tugsbayasgalan replied:
done

tugsbayasgalan committed:
Update on "Add utility methods to find an upgrader"


Differential Revision: [D33198156](https://our.internmc.facebook.com/intern/diff/D33198156)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198156).

tugsbayasgalan committed:
Update on "Add utility methods to find an upgrader"


Differential Revision: [D33198156](https://our.internmc.facebook.com/intern/diff/D33198156)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198156).
